### PR TITLE
fix(dashboard): roll back a mid-replay attach drop so idle reaping can reclaim the PTY

### LINKED ISSUE #110849: PtySession.attach() commits attached state before the replay write — a client dropping during attach leaves the PTY permanently unreapable
### Summary

`PtySession.attach()` commits the attached state *before* it writes the replay buffer to the socket, and the caller (`pty_ws`) performs that call **outside** any `try/finally`. A client that drops during the attach replay therefore raises out of the handler with the session already marked `attached=True, last_detached_at=None` and never reaches `PTY_REGISTRY.detach()`.

That state is the one state `reap_idle()` can never reclaim:

* `not s.alive` is `False` — the PTY child is perfectly alive;
* the second clause requires `not s.attached` **and** `s.last_detached_at is not None` — both are false.

`_reap_one_idle_or_raise()` has the same `not attached` requirement, and the 30-minute TTL is anchored to `last_detached_at`, which stays `None` forever. The PTY process group (`node ui-tui/dist/entry.js` + its `python -m tui_gateway.entry` child) then lives indefinitely, holding a tty, the PTY master fd, its loopback sockets and the session's work.

This is an independent entry into "attached forever" — it needs no reverse proxy, no disabled WS ping, and no SIGHUP cascade; the client only has to disappear between `attach` bookkeeping and the first byte of the replay.

### Where (verified on `main` @ a89c1e11)

`hermes_cli/pty_session.py` (attach, ~L98-L114):

```python
if self._ws is not ws:
    await _close_ws(self._ws, WS_CLOSE_SUPERSEDED)
self._ws = ws
self._attach_generation += 1
self.attached = True
self.last_detached_at = None
if snap := self.buffer.snapshot():
    await ws.send_bytes(snap)          # <- raises here, state already committed
```

`hermes_cli/web_routers/chat_ws.py` (L516) — the call site is outside the `try` that owns the detach; `PTY_REGISTRY.detach()` only runs in the writer loop's `finally` further down:

```python
if not await session.attach(ws, force_redraw=not _created):
    await _close_stalled_pty_input(ws, path="keepalive-redraw")
    PTY_REGISTRY.detach(attach_token, ws)
    return
```

### Evidence

Dashboard server log (3 occurrences, same shape), client-side `/api/pty` socket:

```
chat_ws.py:516 in pty_ws
pty_session.py:111 in attach
    await ws.send_bytes(snap)
starlette/websockets.py:89 in send -> raise WebSocketDisconnect(code=1006)
...
uvicorn.protocols.utils.ClientDisconnected
```

Field accumulation on one `hermes dashboard` with a single real terminal (`who` → console + one SSH session):

* peak snapshot: **18** `node ui-tui/dist/entry.js` and **15** `python -m tui_gateway.entry` processes, **15** ttys held, **15** PTY master fds in the dashboard process;
* all ESTABLISHED sockets on the dashboard port were held by the dashboard process itself (no browser clients at all) — i.e. no live viewer anywhere, yet every session stayed attached;
* one orphaned `ui-tui` node (no `tui_gateway` child) survived from the same day five days earlier and was still present in every later snapshot;
* a rescan 45 minutes later showed the total fluctuating (8 nodes / 7 children) but the oldest orphan unchanged — the leak is cumulative, not a one-off.

### Impact

Unbounded process / PTY / fd leak per dropped client, plus indirectly a stuck session lease: a stranded `tui_gateway` with an unfinished turn keeps refreshing its `session_turn_leases` row, which is invisible to watchdog checks keyed on "lease expired" or "owner pid dead" (the expiry is fresh and the pid is alive).

### Proposed fix (any of)

1. In `attach()`, do not commit `_ws`/`attached`/`last_detached_at` until the replay buffer has been delivered — send first, then record the attachment (or roll back in `except`).
2. In `pty_ws`, wrap the `session.attach(...)` call in `try/except` and `PTY_REGISTRY.detach(attach_token, ws)` before re-raising (or fold the attach into the writer loop's `try/finally`).
3. In `reap_idle()`, add a criterion independent of `attached` (e.g. an attach timestamp / socket-liveness probe) so a session whose viewer is gone is eventually reclaimed even if the detach was never observed.

Regression test: a `FakeWS` whose `send_bytes` raises → after the failed attach the session must still be reclaimable (`attached is False` or `last_detached_at is not None`). The existing suite pins the opposite invariant (`tests/test_pty_session.py` — "attached → not reapable") but does not cover an aborted attach.

### Related

* #110164 — orphaned `ui-tui` nodes with dead gateway children; the attach-abort path above is another, proxy-independent source of the same accumulation.
* #96418 — attached session never reaped when the loopback bind disables the WS ping (external trigger); this report is the in-process trigger and does not require the ping to be off.
* #106400 — `reap_idle()` `pop(key)` without default; fixed on `main` (a89c1e11) and **not** the same path as this report.
* PR #110168 — keeps the embedded dashboard TUI alive on SIGHUP; unrelated to the stale-attached state.

### PR DESCRIPTION
**Pull request by liuhao1024**
## What does this PR do?

`PtySession.attach()` commits the attached state (`_ws`, `attached = True`, `last_detached_at = None`) **before** it writes the replay buffer to the socket, and `pty_ws` performs that call outside the `try` whose `finally` owns `PTY_REGISTRY.detach()`. A client that drops during the attach replay raises out of the handler with the session already marked attached — the one state `reap_idle()` can never reclaim (it requires `not alive`, or detached with a timestamp). The PTY process group (`node ui-tui` + its `tui_gateway` child, tty, master fd, loopback sockets) then lives indefinitely; the issue documents cumulative accumulation on a real dashboard.

This PR makes `attach()` undo the committed attach when the replay write fails: the session reads detached+idle, the TTL reaper can reclaim it, and the returned `False` routes the caller into its existing close-and-detach branch. The rollback is safe to do unconditionally here because at replay time no viewer ever owned the session — unlike a drain-loop failure, there is no live viewer to misclassify.

Scope note: the reported traces (`chat_ws.py:516` → `pty_session.py:111` `await ws.send_bytes(snap)`) are all on the attach-replay path, and that is what this PR closes. The `_drain` forwarding path raised in [this comment](https://github.com/NousResearch/hermes-agent/issues/110849#issuecomment-5664276368) can enter the same unreapable state, but distinguishing a dead client from a transient send failure there (a wide `except` would detach a live viewer) is a behavior decision I have deliberately left out of this focused fix.

## Related Issue

Fixes #110849

## Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)

## Changes Made

- `hermes_cli/pty_session.py` — wrap the replay `send_bytes` in `attach()` with a rollback to the detached/idle form (`_ws = None`, `attached = False`, `last_detached_at = now`) on failure and return `False`, so the caller's existing cleanup branch runs and `reap_idle()` can reclaim the session
- `tests/hermes_cli/test_pty_session.py` — two regression tests: a mid-replay drop rolls back to detached+idle while the PTY child stays alive, and a subsequent replacement socket can still attach and receives the replay

## How to Test

1. `python -m pytest tests/hermes_cli/test_pty_session.py -q` — should pass: 16 passed (14 pre-existing + 2 new)
2. `python -m pytest tests/hermes_cli/test_pty_keepalive_ws.py tests/hermes_cli/test_web_server_pty_reconnect.py tests/hermes_cli/test_web_server_pty_idle_backoff.py -q` — should pass: 10 passed
3. Observed result: both new tests fail on unpatched `main` (the session stays `attached=True`, `last_detached_at=None`, i.e. unreapable) and pass with this change; no pre-existing test regressed.

### Code

- [x] I've read the [Contributing Guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md)
- [x] My commit messages follow [Conventional Commits](https://www.conventionalcommits.org/) (`fix(scope):`, `feat(scope):`, etc.)
- [x] I searched for [existing PRs](https://github.com/NousResearch/hermes-agent/pulls) to make sure this isn't a duplicate
- [x] My PR contains **only** changes related to this fix/feature (no unrelated commits)
- [x] I've run `pytest tests/ -q` and all tests pass
- [x] I've added tests for my changes (required for bug fixes, strongly encouraged for features)
- [x] I've tested on my platform: macOS 15.4 (arm64)

### Documentation & Housekeeping

- [x] I've updated relevant documentation (README, `docs/`, docstrings) — or N/A
- [x] I've updated `cli-config.yaml.example` if I added/changed config keys — or N/A
- [x] I've updated `CONTRIBUTING.md` or `AGENTS.md` if I changed architecture or workflows — or N/A
- [x] I've considered cross-platform impact (Windows, macOS) per the [compatibility guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md#cross-platform-compatibility) — or N/A
- [x] I've updated tool descriptions/schemas if I changed tool behavior — or N/A

## For New Skills

N/A

## Screenshots / Logs

Not applicable (pure Python change with unit-test coverage).

### CODE REVIEW

**Comment by alt-glitch** [comment]
> *This was generated by AI during triage.*

Related: #110859 is a competing fix for #110849 filed at the same time. That PR also detaches on `_drain` send failure and re-raises; this PR rolls back the attach and returns False. Maintainers should pick one mechanism.

**Comment by liuhao1024** [comment]
Thanks for the cross-reference. For maintainers weighing the two mechanisms, the scope difference here is intentional:

- This PR fixes the attach/replay path only, where rollback is zero-harm: at replay time the incoming viewer has not taken ownership of the session yet, so restoring the pre-attach state simply hands it back to the existing idle reaper. It also returns `False` instead of raising, so the ws handler exits through its existing cleanup branch rather than an exception path (the `attach()` call site sits outside the handler `try`).
- #110859 additionally detaches on live `_drain` send failures, which also covers the sibling unreapable case, but it replaces `_drain`'s current `keep buffering` tolerance — a transient send error on a live viewer would now mark the session detached and reapable while that viewer is still attached. Whether that trade-off is wanted (or needs an extra guard) looks like a maintainer call, so I kept this PR narrow and documented the boundary in its description.

**Comment by kyssta-exe** [comment]
## Review: roll back attach on mid-replay drop

**Summary:** Focused fix for #110849: `PtySession.attach()` now undoes its committed attach state (`_ws=None`, `attached=False`, fresh `last_detached_at`) when the replay-buffer write fails, returning `False` so the caller's existing cleanup branch runs and `reap_idle()` can reclaim the PTY group. Two regression tests verified to fail on unpatched main.

**What changed**
- `hermes_cli/pty_session.py`: try/except around replay `send_bytes` in `attach()` with explanatory comment; returns `False` instead of propagating.
- `tests/hermes_cli/test_pty_session.py`: `DroppedClientWS` harness + rollback test and replacement-socket test.

**Strengths**
- The comment explains *why* the rollback must live here (handler unwinds before the writer-loop finally) — future readers won't "simplify" it away.
- Returning `False` keeps the failure on the normal control-flow path rather than exception-driven; replacement attach proven to still get the full replay.

**Findings**
- **Coordination needed:** #110859 and #110863 fix the same issue on the same file. This PR deliberately leaves the `_drain` live-send path open (stated in the body); #110863 closes it with generation scoping. Only one should land — if this one, the drain gap should become a tracked follow-up.
- The rollback open-codes `detach()`'s three assignments instead of calling `self.detach(ws)` — at replay time they're equivalent (no viewer could own the session yet), but a shared helper would prevent future drift. Non-blocking polish.

**Verdict:** Looks good to merge, subject to the dedupe decision with #110859/#110863.

*Reviewed using Hermes-Agent*

**Comment by liuhao1024** [comment]
Thanks for the thorough review — the pointer to #110863's generation-scoping approach is useful context for maintainers.

Adopted the non-blocking polish in dac208ea: the rollback now calls `self.detach(ws)` instead of duplicating its three assignments. At the failure point `_ws` is still the failing socket we just committed, so `detach()`'s superseded-socket guard passes and the call is a drop-in equivalent (16/16 in `tests/hermes_cli/test_pty_session.py`, including the rollback and replacement-socket regressions).

On the dedupe question I'll leave the pick to maintainers, as noted above — happy to file the `_drain` gap as a tracked follow-up if this mechanism is chosen.

**Comment by liuhao1024** [comment]
Housekeeping note: this fix has landed on `main` in a broader form via teknium1's merged #111357 (`fix(dashboard): dropped terminal viewers no longer pin sessions`), which reworks the same `PtySession.attach()` / `_drain()` paths in `hermes_cli/pty_session.py` plus the chat_ws spawn side, and the anchor issue #110849 is closed as completed. It should be safe to close this PR as already applied.
